scipy/sparse/construct.py

```diff
@@ -665,6 +729,10 @@ def block_diag(mats, format=None, dtype=None):
            [0, 0, 0, 7]])
 
     """
+    
+    if all([mat.shape == mats[-1].shape for mat in mats[:-1]] ):
```

Comment:
You can do `all([...]) -> all(...)` to make it a generator expression and avoid constructing the list.